keps/0024-parameter-enhancement.md

```diff
@@ -0,0 +1,146 @@
+owners:
+  - "@nfnt"
+creation-date: 2020-02-21
+last-updated: 2020-02-21
+status: provisional
+---
+
+# Enhanced Operator Parameters
+
+## Table of Contents
+
+- [Enhanced Operator Parameters](#enhanced-operator-parameters)
+  - [Table of Contents](#table-of-contents)
+  - [Summary](#summary)
+  - [Motivation](#motivation)
+    - [Goals](#goals)
+    - [Non-Goals](#non-goals)
+  - [Proposal](#proposal)
+    - [Support Optional Parameter Types](#support-optional-parameter-types)
+    - [Provide Parameter Validation](#provide-parameter-validation)
+    - [Alternative Considered: Convert YAML in string to dictionaries](#alternative-considered-convert-yaml-in-string-to-dictionaries)
+    - [Alternative Considered: Allow Arbitrary Parameter Values](#alternative-considered-allow-arbitrary-parameter-values)
+  - [Implementation History](#implementation-history)
+
+## Summary
+
+This KEP aims to improve the use-cases that can be covered by parameters. By supporting additional types, templates can access lists and dictionaries. Furthermore, validation can be implemented for known types.
+
+## Motivation
+
+Currently only string parameters are supported by KUDO. This limits the possible use-cases of parameters. Various use cases benefit from having parameters that are lists or dictionaries. While some of this functionality can be implemented in Go templates using [Sprig functions][1], this is unintuitive.
+Having parameter types creates the possibility for parameter validation.
+
+### Goals
+
+- Provide operator developers with more versatile parameters
+- Allow the definition of validation functions for parameters
+
+### Non-Goals
+
+- Change how parameters are defined
+
+## Proposal
+
+### Support Optional Parameter Types
+
+By adding an optional `type` field to the list of parameters, parameter types can be specified. If this field isn't present, it defaults to the `string` type. Other types like `list`, `dict` or `integer` are possible. KUDO's renderer uses the type information to convert the string value to the respective type. For example, for the `dict` type, the input value would be YAML that is unmarshalled to a `map[string]interface{}`. Go's template engine allows to access and iterate list and dictionary items.
+
+Consider a cluster that has nodes in multiple regions. We want to ensure that a nginx deployment has 2 replicas in one region and 1 replica in the other one. Furthermore they should support both HTTP as well as HTTPS, hence need a list of port definitions. These parameters are described by `list` and `dict` types:
+
+```yaml
+parameters:
+  - name: ports
+    description: Ports
+    type: list
+    default: [ 80, 443 ]
+  - name: topology
+    description: Node topology
+    type: dict
+    default: |
+      - region: us-east-1:
+        replicas: 2
+      - region: us-west-1:
+        replicas: 1
+```
+
+The deployment template then iterates over the items of the respective parameters:
+
+```yaml
+{{ $ports := .Params.ports }}
```

Comment:
Something to think about: custom variables (other than the special unnamed variable `$`) are currently not supported by the package validator.

Reply:
Is there a reason for that? I'd consider that a bug in the package validator :)
Custom variables become very important when working with lists and dictionaries.

Reply:
Yes, the problem is that arbitrary variables are much closer to a turing-complete language, which is rather hard to validate using a static checker :-)
See https://github.com/kudobuilder/kudo/pull/1255#issue-360463222

Reply:
What are we validating there anyways? If a parameter is used in a template that isn't set, rendering will fail (see https://goplay.space/#d3dyQyolK29). Hence, validation should do just that. What's it validation additionally that needs us deal with these properties?

Reply:
Just ran some tests. I didn't run into issues with the package validator when using variables.
Changed `first-operator` to define `$replicas := .Params.replicas` and used it without issues.

Reply:
What I meant by "not supported" is that that the validator will falsely report some params as unused if they are only referenced indirectly via a custom variable, other than `$`.